rust_src/src/character.rs

```diff
@@ -1,9 +1,42 @@
 //! Operations on characters.
 
+use libc::{c_uchar, ptrdiff_t};
 use lisp::LispObject;
 use multibyte::{make_char_multibyte, raw_byte_from_codepoint_safe, MAX_CHAR};
 use remacs_macros::lisp_fn;
 use remacs_sys::EmacsInt;
+use threads::ThreadState;
+
+/// True iff byte starts a character in a multibyte form.
+///
+/// Same as the `CHAR_READ_P` macro.
+#[inline]
+pub fn char_read_p(byte: c_uchar) -> bool {
+    (byte & 0xC0) != 0x80
+}
+
+/// Decrement the buffer byte position POS_BYTE of the current buffer
+/// to the previous character boundary. No range checking of POS.
+///
+/// Can be used instead of the `DEC_POS` macro.
+#[inline]
+pub unsafe fn dec_pos(pos_byte: &ptrdiff_t) -> ptrdiff_t {
```

Comment:
The C code from src/char.h

    #define DEC_POS(pos_byte)			\
     do {						\
        unsigned char *chp;				\
    	    					\
        pos_byte--;					\
        if (pos_byte < GPT_BYTE)			\
          chp = BEG_ADDR + pos_byte - BEG_BYTE;	\
        else					\
          chp = BEG_ADDR + GAP_SIZE + pos_byte - BEG_BYTE; \
        while (!CHAR_HEAD_P (*chp))			\
          {						\
	      chp--;					\
	      pos_byte--;				\
          }						\
      } while (false)


Notice that `pos_byte` is decremented. This is not what the Rust code does. the value of `pos_byte` is copied. This might lead to subtle bugs.

Reply:
I originally implemented it that way, but as far as I could tell the macro was never used for side effects (i.e. for pointer arithmetic). That's why I'd decided to implement it in a more idiomatic way instead of mutating a reference. The rust compiler should be able to optimize the copy away, so it's purely a matter of personal preference ([example](https://godbolt.org/g/GTyMu6)).

If you'd prefer to stay true to the original C code I could of course change the function to mutate the reference!

Reply:
I just looked at the code and you are right `DEC_POS`. They just modify the pos in place because they cannot return.

Reply:
Next question. Does the whole function have to be unsafe? The rest of remacs tries to limit the scope of that.

Reply:
Yeah that's a weird one. The way it's used in the C codebase it should never panic, but it doesn't do any bounds checking so an incorrect `pos_byte` argument could lead to segfaults.

Reply:
If `pos_byte` is not being modified there is no reason to pass a reference to an integer. Just pass `pos_byte` by value.